### PR TITLE
✨ feature: List artifacts + inspect individual artifacts

### PR DESCRIPTION
**Pull request by queer**
image

image

### CODE REVIEW

File: cmd/helpdocs/artifact.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# artifact command help docs
+
+TODO
```
Comment (queer) [inline cmd/helpdocs/artifact.md]: Unsure how to best fill this out

File: internal/api/artifact.go [hunk 2]
```diff
@@ -1,10 +1,87 @@
...
+		}
+
+		out = append(out, nextRes...)
+	}
+
+	return out, nil
+}
+
+func GetArtifact(client graphql.Client, orgID string, artifactID string) (*Artifact, error) {
+	// TODO(amy): The backend needs to permit service accounts to run this query!
```
Comment (queer) [inline internal/api/artifact.go:73]: I haaaaaaaaate it and should/probably-will follow up w/ a PR for this

File: internal/api/genqlient.graphql [hunk 3]
```diff
@@ -113,3 +113,39 @@ query containerRepository($orgId: ID!, $artifactId: ID!, $input: ContainerReposi
...
+      data
+      artifactDefinition {
+        id
+        name
+        schema
+        url
+      }
+    }
+  }
+}
```
Comment (queer) [inline internal/api/genqlient.graphql:151]: Awkward and I hate it, but I can't pass a `nil` string when the query is `getAllArtifacts($orgId: ID!, $after: String)`? Maybe I'm just not used to Go's type system...

File: cmd/artifact.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+var listArtifactsCommand = &cobra.Command{
+	Use:     "list",
+	Aliases: []string{"l", "ls"},
+	Short:   "List all artifacts",
+	RunE:    runListArtifacts,
+}
+
+var artifactInfoCommand = &cobra.Command{
```
Comment (coryodaniel) [inline cmd/artifact.go:29]: I’d like to keep the CLI verb oriented get/list/create

File: cmd/artifact.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+	Use:     "info <artifact id>",
+	Aliases: []string{"i"},
+	Short:   "Get information about an artifact",
```
Comment (coryodaniel) [inline cmd/artifact.go:32]: I’d like to keep the CLI verb oriented get/list/create

File: cmd/artifact.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+func init() {
+	rootCmd.AddCommand(artifactCommand)
+	artifactCommand.AddCommand(listArtifactsCommand)
+	artifactCommand.AddCommand(artifactInfoCommand)
+}
+
+func runListArtifacts(cmd *cobra.Command, args []string) error {
+	c := config.Get()
+	client := api.NewClient(c.URL, c.APIKey)
+	artifacts, err := api.GetAllArtifacts(client, c.OrgID)
```
Comment (coryodaniel) [inline cmd/artifact.go:46]: Push all logic and presentation down into commands/

File: cmd/artifact.go [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("Artifacts:")
```
Comment (coryodaniel) [inline cmd/artifact.go:52]: Use a bubble tea table for presenting a table of items

https://github.com/Evertras/bubble-table

File: cmd/artifact.go [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+		fmt.Println(msg)
+	}
+
+	return nil
+}
+
+func runArtifactInfo(cmd *cobra.Command, args []string) error {
+	c := config.Get()
+	client := api.NewClient(c.URL, c.APIKey)
+	artifact, err := api.GetArtifact(client, c.OrgID, args[0])
```
Comment (coryodaniel) [inline cmd/artifact.go:66]: Same notes as above regarding command and bubble tea. We are trying to keep the cobra command as soon as possible and encapsulate all of the display logic.

File: internal/api/artifact.go [hunk 9]
```diff
@@ -1,10 +1,87 @@
...
+)
+
 type Artifact struct {
-	Name string
-	ID   string
+	Name               string
+	ID                 string
+	Type               string
+	ArtifactDefinition ArtifactDefinition
+	Data               interface{}
```
Comment (coryodaniel) [inline internal/api/artifact.go:15]: I believe this is a map[string]interface{} since it’s JSON data

File: internal/api/artifact.go [hunk 10]
```diff
@@ -1,10 +1,87 @@
...
 }
 
 type ArtifactDefinition struct {
+	ID   string
 	Name string
+	Url  string
+}
+
+func GetAllArtifacts(client graphql.Client, orgID string) ([]Artifact, error) {
+	out, err := getAllArtifactsWithPagination(context.Background(), client, orgID, nil)
```
Comment (coryodaniel) [inline internal/api/artifact.go:26]: Call it what it is “artifacts” instead of “out”

File: internal/api/artifact.go (hunk 2 above)
Comment (coryodaniel) [reply internal/api/artifact.go:73]: There’s nothing more permanent than a temporary solution ;)

File: internal/api/credential.go [hunk 11]
```diff
@@ -8,10 +8,10 @@ import (
 )
 
 var credentialArtifactDefinitions = []*ArtifactDefinition{
-	{"massdriver/aws-iam-role"},
-	{"massdriver/azure-service-principal"},
-	{"massdriver/gcp-service-account"},
-	{"massdriver/kubernetes-cluster"},
+	{"massdriver/aws-iam-role", "AWS IAM Role", ""},
```
Comment (coryodaniel) [inline internal/api/credential.go:11]: Why are these changing? This is specifically for integrating credentials definitions not _all artifacts_ … we also have a mediocre inflector that does this. If we need it elsewhere we should bring it up from the arttdef table UI component as a strict function Titleize()

but also still curious why credential struct being modified wrt listing _all_ artifacts

File: internal/api/genqlient.graphql [hunk 12]
```diff
@@ -113,3 +113,39 @@ query containerRepository($orgId: ID!, $artifactId: ID!, $input: ContainerReposi
...
 }
+
+query getAllArtifacts($organizationId: ID!) {
+  artifacts(organizationId: $organizationId, input: {}) {
+    next
+    items {
+      id
+      name
+      type
+      data
```
Comment (coryodaniel) [inline internal/api/genqlient.graphql:124]: Suspect re: data here. @xpositivityx are we auditing this field access?

File: internal/api/genqlient.graphql (hunk 3 above)
Comment (coryodaniel) [reply internal/api/genqlient.graphql:151]: Not sure I follow. Is the backend typed not nil? I see a “!” In the Gql

you’ll need a string pointer to pass nil or pass a “nilish” string “”

File: internal/commands/artifacts.go [hunk 13]
```diff
@@ -0,0 +1,14 @@
+package commands
+
+import (
+	"github.com/Khan/genqlient/graphql"
+	"github.com/massdriver-cloud/mass/internal/api"
+)
+
+func GetAllArtifacts(client graphql.Client, orgID string, name string) ([]api.Artifact, error) {
```
Comment (coryodaniel) [inline internal/commands/artifacts.go:8]: ListArtifacts to stay consistent 

logic in your cobra commands should be moved here + tests

Goal is to make commands/ testable versions of the cobra commands but without any terminal I/O flag parsing outputting to files etc

**Review by coryodaniel — CHANGES_REQUESTED** [review]
Just some changes around consistentency

**Pull request closed without merging**